ocaml/testsuite/tests/typing-layouts/error_message_attr.ml

```diff
@@ -116,3 +116,130 @@ Error: Bad layout annotation:
+   Seperate implementation from when it's applied to layout annotations. *)
+
+(* Needs a string body *)
+let f (x : bool) = (x : int)[@error_message]
+[%%expect{|
+Line 1, characters 28-44:
+1 | let f (x : bool) = (x : int)[@error_message]
+                                ^^^^^^^^^^^^^^^^
+Warning 47 [attribute-payload]: illegal payload for attribute 'error_message'.
+Error_message attribute expects a string argument
+Line 1, characters 20-21:
+1 | let f (x : bool) = (x : int)[@error_message]
+                        ^
+Error: This expression has type bool but an expression was expected of type
+         int
+|}]
+
+(* Can only be applied once *)
+let f (x : bool) = (x : int)[@error_message "A"][@error_message "B"]
+[%%expect{|
+Line 1, characters 48-68:
+1 | let f (x : bool) = (x : int)[@error_message "A"][@error_message "B"]
+                                                    ^^^^^^^^^^^^^^^^^^^^
+Warning 47 [attribute-payload]: illegal payload for attribute 'error_message'.
+More than one error_message attribute present. All of them will be ignored.
+Line 1, characters 20-21:
+1 | let f (x : bool) = (x : int)[@error_message "A"][@error_message "B"]
+                        ^
+Error: This expression has type bool but an expression was expected of type
+         int
+|}]
+
+(* Simple test case *)
+let f (x : bool) = (x : int)[@error_message "custom message"]
+[%%expect{|
+Line 1, characters 20-21:
+1 | let f (x : bool) = (x : int)[@error_message "custom message"]
+                        ^
+Error: This expression has type bool but an expression was expected of type
+         int
+       because [@error_message]: custom message
+|}]
+
+(* Doesn't work when the type mismatch happens later. This differ from
+   the layout annotation case. *)
+let f x: bool = (x : int)[@error_message "custom message"]
+[%%expect{|
+Line 1, characters 16-25:
+1 | let f x: bool = (x : int)[@error_message "custom message"]
+                    ^^^^^^^^^
+Error: This expression has type int but an expression was expected of type
+         bool
+|}]
+
+(* Doesn't apply when the type error is from elsewhere within the expression *)
+let g (x : int) = x
+let f (x : bool) = (let y = false in g y : int)[@error_message "custom message"]
+[%%expect{|
+val g : int -> int = <fun>
+Line 2, characters 39-40:
+2 | let f (x : bool) = (let y = false in g y : int)[@error_message "custom message"]
+                                           ^
+Error: This expression has type bool but an expression was expected of type
+         int
+|}]
+
+(* Can be used to enforce layouts but not great *)
+let f (x : string) = (x : (_ : immediate))[@error_message "custom message"]
+[%%expect{|
+Line 1, characters 22-23:
+1 | let f (x : string) = (x : (_ : immediate))[@error_message "custom message"]
+                          ^
+Error: This expression has type string but an expression was expected of type
+         ('a : immediate)
+       because [@error_message]: custom message
```

Comment:
As with layouts, it would be cool to put this at the end. But that might be too hard.

Reply:
I agree. Looked a bit more into this and don't see a simple way of moving the message to the end.